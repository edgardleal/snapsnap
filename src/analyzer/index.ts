export * from './analyzer';
export * from './nock';
