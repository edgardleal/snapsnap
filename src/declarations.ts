/* eslint-disable @typescript-eslint/no-namespace, no-redeclare */

export {};

declare global {
  namespace jest {
    interface It {
      snap: It;
    }
  }
}
