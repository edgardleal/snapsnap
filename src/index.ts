export * from './analyzer';
export * from './lock';
export * from './lock-manager';

export * as integrations from './integrations';
