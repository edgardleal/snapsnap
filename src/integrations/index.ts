export * as jest from './jest';
