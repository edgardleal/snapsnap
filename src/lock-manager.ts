export class LockManager {
  private promise   : Promise<void>;

  // eslint-disable-next-line @typescript-eslint/ban-types
  private resolver? : Function;

  public constructor() {
    this.promise = new Promise(
      (resolve) => {
        this.resolver = resolve;
      },
    );
  }

  public release(): void {
    if (!this.resolver) {
      throw new Error('Unknown resolver');
    }

    this.resolver();

    delete this.resolver;
  }

  public get(): Promise<void> {
    return this.promise;
  }
}

