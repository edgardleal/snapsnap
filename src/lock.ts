import AsyncLock from 'async-lock';

import { LockManager } from './lock-manager';


export class Locker {
  private static readonly LOCK_NAME = 'snapshot-mock-lock';

  private lock = new AsyncLock();

  public acquire(): Promise<LockManager> {
    return new Promise(
      (resolve, reject) => {
        try {
          this.lock.acquire(
            Locker.LOCK_NAME,
            async () => {
              const lm = new LockManager();

              resolve(lm); // ready to execute

              await lm.get(); // wait until executed
            },
          ).catch((err) => reject(err));
        } catch (err) {
          reject(err);
        }
      },
    );
  }
}

export const globalLock = new Locker();
