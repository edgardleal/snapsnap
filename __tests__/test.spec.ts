import fetch from 'node-fetch';

import '../src/declarations';

it.snap(
  'should deal with error responses',
  async () => {
    const response = await fetch('https://postman-echo.com/missinig-incorrect-url');

    expect(response.status).toEqual(404);
  },
);

it.snap(
  'should use Nock Back cache to resolve queries',
  async () => {
    const response = await fetch('https://postman-echo.com.broken:4443/get?foo1=bar1&foo2=bar2');

    expect(await response.json()).toMatchSnapshot();
  },
);

it.snap(
  'should resolve queries',
  async () => {
    const url = 'https://postman-echo.com/get?another=query';
    const response = await fetch(url);

    const data = await response.json();

    expect(data.url).toEqual(url);
    expect(data.args.another).toEqual('query');
    expect(data.headers.host).toEqual('postman-echo.com');

    expect(data).toMatchSnapshot();
  },
);
