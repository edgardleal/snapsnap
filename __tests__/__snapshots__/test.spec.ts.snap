// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should resolve another query 1`] = `
Object {
  "args": Object {
    "another": "query",
  },
  "headers": Object {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "host": "postman-echo.com",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
    "x-amzn-trace-id": "Root=1-629d65c9-32fd19d246c68ccb20649c43",
    "x-forwarded-port": "443",
    "x-forwarded-proto": "https",
  },
  "url": "https://postman-echo.com/get?another=query",
}
`;

exports[`should resolve queries 1`] = `
Object {
  "args": Object {
    "another": "query",
  },
  "headers": Object {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "host": "postman-echo.com",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
    "x-amzn-trace-id": "Root=1-629d661c-5220cece0d14247b1f34d5dc",
    "x-forwarded-port": "443",
    "x-forwarded-proto": "https",
  },
  "url": "https://postman-echo.com/get?another=query",
}
`;

exports[`should use Nock Back cache to resolve queries 1`] = `
Object {
  "args": Object {
    "foo1": "bar1",
    "foo2": "bar2",
  },
  "headers": Object {
    "accept": "*/*",
    "accept-encoding": "gzip,deflate",
    "host": "postman-echo.com",
    "user-agent": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)",
    "x-forwarded-port": "443",
    "x-forwarded-proto": "https",
  },
  "url": "https://postman-echo.com/get?foo1=bar1&foo2=bar2",
}
`;
